sui_programmability/framework/sources/TestScenario.move

```diff
@@ -124,26 +122,74 @@ module Sui::TestScenario {
     /// Remove the object of type `T` from the inventory of the current tx sender in `scenario`.
     /// An object is in the sender's inventory if:
     /// - The object is in the global event log
-    /// - The sender owns the object, or the object is immutable
+    /// - The sender owns the object
     /// - If the object was previously removed, it was subsequently replaced via a call to `return_object`.
     /// Aborts if there is no object of type `T` in the inventory of the tx sender
     /// Aborts if there is >1 object of type `T` in the inventory of the tx sender--this function
     /// only succeeds when the object to choose is unambiguous. In cases where there are multiple `T`'s,
     /// the caller should resolve the ambiguity by using `take_object_by_id`.
     public fun take_object<T: key>(scenario: &mut Scenario): T {
-        let sender = sender(scenario);
-        remove_unique_object(scenario, sender)
+        let signer_address = sender(scenario);
+        let objects: vector<T> = get_account_owned_inventory<T>(
+            signer_address,
+            last_tx_start_index(scenario)
+        );
+        remove_unique_object_from_inventory(scenario, objects)
+    }
+
+    /// Similar to take_object, but only return objects that are immutable with type `T`.
+    /// In this case, the sender is irrelevant.
+    /// Returns a wrapper that only supports a `borrow` API to get the read-only reference.
+    public fun take_immutable_object<T: key>(scenario: &mut Scenario): ImmutableWrapper<T> {
```

Comment:
Can't remember if this was discussed/dismissed, but I *think* we could return a `&T` directly here by introducing a magic native function `box<T>(t: T): &T`. I believe the VM heap allocates refs, so this is expressible via the `Value` API.

Reply:
@tnowacki Do you know which API may be able to do this?

Reply:
I took a look and I think this is possible to do, though not with the public API's of `VMValue`. The code I am looking at is:

```
  pub fn signer_reference(x: AccountAddress) -> Self {
        Self(ValueImpl::ContainerRef(ContainerRef::Local(
            Container::signer(x),
        )))
    }
```

If we exposed `value_reference_for_testing(v: Value) -> Self` that can be used to implement `box<T>` and the mut variant in Move.

Reply:
You could sort of get away with this... but you would definitely need a bunch of dynamic checks internally. Basically you need to make sure the refcount for the mutable case is 0 before giving a new borrow.
Similarly for the immutable borrow you would need to know there are no mutable references out there.
Basically similar to the reference safety rules, but dynamic (which is what used to happen for global storage before the `acquires` thing. 

I was hoping to get rid of these too FWIW, since the checks aren't *really* needed because of the `acquire` rules. It was just too much for this PR.

Conceptually though, there isn't any real reason you couldn't add an `Rc` type to Move, or even just `leak<T:drop>(T): &mut T` (though the second one would require some borrow checker changes)